File: 658-FindKClosestElements/658-FindKClosestElements.py
class Solution(object):
    def findClosestElements(self, arr, k, x):
        """
        :type arr: List[int]
        :type k: int
        :type x: int
        :rtype: List[int]
        """
        heap = []
        
        for i in range(k):
            heapq.heappush(heap, (-abs(arr[i] - x), -arr[i]))
        
        for i in range(k, len(arr)):
            diff = abs(arr[i] - x)
            if diff < -heap[0][0] or (diff == -heap[0][0] and arr[i] < -heap[0][1]):
                heapq.heappushpop(heap, (-diff, -arr[i]))
        
        result = sorted(-num for _, num in heap)
        return result

File: 658-FindKClosestElements/658-FindKClosestElements.java
class Solution {
    public List<Integer> findClosestElements(int[] arr, int k, int x) {      
        int start = 0, end = 0;
        int minDiff = Integer.MAX_VALUE;
        int curDiff = 0;

        for (end = 0; end < k; end++) {
            curDiff += Math.abs(x - arr[end]);
        }

        minDiff = curDiff;
        int bestStart = 0;

        while (end < arr.length) {
            curDiff += Math.abs(x - arr[end]);
            curDiff -= Math.abs(x - arr[start]);

            if (curDiff < minDiff) {
                bestStart = start + 1; 
                minDiff = curDiff;
            }

            start++;
            end++;
        }

        List<Integer> result = new ArrayList<>();
        for (int i = bestStart; i < bestStart + k; i++) {
            result.add(arr[i]);
        }
        
        return result;
    }
}